lib/portage/dbapi/bintree.py

```diff
@@ -2199,6 +2173,41 @@ def _allocate_filename_multi(self, cpv):
             else:
                 return filename
 
+    def _allocate_build_id(self, cpv):
+        # First, get the max build_id found when _populate was
+        # called.
+        max_build_id = self._max_build_id(cpv)
+
+        # A new package may have been added concurrently since the
+        # last _populate call, so use increment build_id until
+        # we locate an unused id.
+        pf = catsplit(cpv)[1]
+        build_id = max_build_id + 1
+
+        try:
+            binpkg_format = cpv.binpkg_format
+        except AttributeError:
+            binpkg_format = self.settings.get(
+                "BINPKG_FORMAT", SUPPORTED_GENTOO_BINPKG_FORMATS[0]
+            )
+
+        if binpkg_format == "xpak":
+            filename_format = "%s-%s.xpak"
+        elif binpkg_format == "gpkg":
+            filename_format = "%s-%s.gpkg.tar"
+        else:
+            raise InvalidBinaryPackageFormat(binpkg_format)
+
+        while True:
+            filename = filename_format % (
+                os.path.join(self.pkgdir, cpv.cp, pf),
+                build_id,
+            )
+            if os.path.exists(filename):
```

Comment:
Can you perhaps create the file here (e.g. as an empty file) to actually avoid races? Using `x` as mode should do the trick.

Reply:
Sure.

Reply:
This will leave an empty file if binpkg creation failed, and increase the build id. Is this ok?

Reply:
This one actually created a lot of invalid binary package warning. I guess I need silence it if the file is empty.

Reply:
Does binpkg creation fail that lot, or are you talking about tests?

Reply:
It is actually because `getname` does not know what status the file is, and try to read the empty placeholder.

* Exists package.
* New package.
* New allocated empty placeholder.
* Remote package.

The placeholder file is implemented, but the logic in a lot of other places has been modified.